docs/proposals/drafts/20211007-pluggable-authenticators.md

```diff
@@ -0,0 +1,143 @@
+
+```
+- name: test-oidc
+  id: 1610b0c3-c509-4592-a256-a1871353dbfa
+  authenticator:
+    type: oidc
+    config:
+      clientID: test
+      clientSecret: xyz
+      issuerCAPath: ./tmp/certs/ca.pem
+      issuerURL: http://127.0.0.1:5556/dex
+      redirectURL: https://localhost:8443/oidc/test-oidc/callback
+      usernameClaim: email
+  rateLimits:
+    - endpoint: "/api/metrics/v1/.+/api/v1/receive"
+      limit: 100
+      window: 1s
+    - endpoint: "/api/logs/v1/.*"
+      limit: 100
+      window: 1s
+```
+
+## 2. Go Plugin Module
+
+In this approach, a generic interface for Authentication providers (Authenticators) is introduced. Authentication providers then implement that interface in a Go plugin. This plugin will be built and loaded at run time by the Observatorium API. How and where to load the plugin module is defined by the tenants configuration file.
+
+This approach is easy to implement, very flexible in terms of adding more authenticators without any additional implementation and integration complexity. However, there are few concerns when it comes to trust and security, since the Observatorium API would consume a pre-built plugin binary as an authenticator.
+
+**Reference**: [Writing Modular Go Programs with Plugins](https://medium.com/learning-the-go-programming-language/writing-modular-go-programs-with-plugins-ec46381ee1a9)
+
+## 3. OAuth2 Providers
+
+Similar to the [generic OAuth2 providers in grafana](https://github.com/grafana/grafana/blob/main/pkg/login/social/social.go), OAuth2 providers can be introduced and injected into the supported OAuth2 alongside to the current implementation of OIDC and mTLS to prevent any disruption for the current adopters. While this approach provides flexibility and can be considered to be an elegant and clean way of introducing more authenticators support, it requires a lot of changes in the Observatorium API.
+
+However, introducing a new authenticator with private or custom implementation would require changing/patching existing Go source files in the Observatorium API. As a result, adopters need to re-apply their patch whenever migrating to a new Observatorium API release with risk to be affected by merge conflicts that must be resolved manually.
+
+## Conclusion
+
+We propose to use the first approach (importing for side-effects), as it appears to be the most flexible approach to follow and requires no additional changes to onboard a new authenticator or additional resources to manage.
+
+It worth mentioning that the new changes must be backward-compatible, in other words, they must not break neither the existing configuration nor the current authentication flows.
+
+## Migration Plan
+
+The current configuration format will not be broken but will be deprecated, so immediate migration is not necessary. Tenants configured to use the old authenticators, namely OIDC, mTLS, and OpenShift, should eventually be migrated to use the new configuration format:
```

Comment:
There's a `:` at the end of this sentence but there is no example following it. Instead there is another section. We should either include an example or turn this into a `.`